src/main/java/com/google/sps/data/Election.java

```diff
@@ -91,4 +71,123 @@ public Entity toEntity() {
+   *     API
+   * @return the new Election object
+   */
+  public Election fromVoterInfoQuery(DatastoreService datastore, JSONObject voterInfoQueryData)
+      throws JSONException {
+    HashSet<Long> contestKeyList = this.getContests();
+    HashSet<Long> propositionKeyList = new HashSet<>();
+    if (voterInfoQueryData.has("contests")) {
+      JSONArray contestListData = voterInfoQueryData.getJSONArray("contests");
+      for (Object contestObject : contestListData) {
+        JSONObject contest = (JSONObject) contestObject;
+
+        long contestEntityKeyId =
+            Contest.fromVoterInfoQuery(datastore, contest).addToDatastore(datastore);
+        contestKeyList.add(contestEntityKeyId);
+      }
+    }
+
+    // TODO(caseyprice): get values for propositions
+
+    return this.withContests(contestKeyList);
+  }
+
+  /**
+   * Checks if an Election object has been populated by the output of a voterInfoQuery call from the
+   * Google Civic Information API.
+   *
+   * @return true if contests and propositions contain elements, false otherwise
+   */
+  public boolean isPopulatedByVoterInfoQuery() {
+    return !getContests().isEmpty() && !getPropositions().isEmpty();
+  }
+
+  /**
+   * Creates a new Election object based on the properties of an Election Entity.
+   *
+   * @param entity the Election Entity to deep copy
+   * @return the new Election object
+   */
+  public static Election fromEntity(Entity entity) {
+    HashSet<Long> contests = new HashSet<>();
+    HashSet<Long> propositions = new HashSet<>();
+    if (entity.getProperty("contests") != null) {
+      contests = (HashSet<Long>) entity.getProperty("contests");
+    }
+
+    if (entity.getProperty("propositions") != null) {
+      propositions = (HashSet<Long>) entity.getProperty("propositions");
+    }
+
+    return Election.builder()
+        .setId((String) entity.getProperty("id"))
+        .setName((String) entity.getProperty("name"))
+        .setDate((String) entity.getProperty("date"))
+        .setScope((String) entity.getProperty("scope"))
+        .setContests(contests)
+        .setPropositions(propositions)
+        .build();
+  }
+
+  /**
+   * Creates a new Entity with properties based on this object's members and stores in the
+   * Datastore.
+   *
+   * @param datastore the DatastoreService to store the new Entity
+   */
+  public long addToDatastore(DatastoreService datastore) {
+    return putInDatastore(datastore, new Entity("Election"));
+  }
+
+  /**
+   * Assigns the given Entity the properties of this Election object and puts it into the given
+   * Datastore instance.
+   *
+   * @param datastore the DatastoreService to store the new Entity
+   */
+  public long putInDatastore(DatastoreService datastore, Entity entity) {
+    /* The "id" of an Election Entity is stored as a property instead of replacing the
+     * Datastore-generated ID because Datastore may accidentally reassign IDs to other
```

Comment:
tbh I think all of this can be avoided if we just fetch all of the data upfront so that we won't need to store the civic API's election ID counterpart anymore, but we can just leave it as is.  (no change required here)